### PR TITLE
AOS-47: Additional Error Specification

### PR DESCRIPTION
**Pull request by xxchan**
rendered ver: https://github.com/xxchan/specs/blob/error/rfcs/47-additional-error-specification.md

### CODE REVIEW

File: rfcs/47-better-error-handing.md [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    - sentinel errors wrapped by `fmt.Errorf` with `%w`: 
+    	- defined in either [go-storage/services/error.go] or `go-service-*/error.go`
+    	- defined as `var ErrSomethingWrong = errors.New("what happened")`
+    	- returned in `fmt.Errorf("%w: %v", services.ErrObjectNotExist, err)`, carrying error messages of original SDK errors
+    - another layer of error `struct`s carrying contextual information: `MetadataUnrecognizedError`, `PairUnsupportedError`, `PairRequiredError`
+    	- defined in [go-storage/services/error.go]
+    	- have `Err` and context fields, don't have `Op`
+    	- returned by constructor methods, where `Err` is set to a sentinel error, which can be viewed as the error `struct`s' classification
+
+![](https://user-images.githubusercontent.com/37948597/116873067-9ce86c80-ac49-11eb-9409-91dcc84da340.png)
```
Comment (Xuanwo) [inline rfcs/47-better-error-handing.md:55]: Please upload images to `rfcs/44`, so that we can link to image like ``

File: rfcs/47-better-error-handing.md [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+  - Their `Err` fields are also confusing.
+    - Will they wrap yet another layer of error `struct`s? 
+    - If it will be the same for all instances, not intended for the caller to fill in, and only accessed via `Unwrap`, we can actually hide the field.
+    - The wrapped errors seem like the super class. It may be more reasonable to wrap other errors instead of being wrapped. But the order does not matter in the error chain.
+
+## Proposal
+
+So I propose the following error handling specification as a supplement of `11-error-handling`:
+
+1. Public APIs SHOULD return top-level errors, which MUST be defined as below:
```
Comment (Xuanwo) [inline rfcs/47-better-error-handing.md:101]: Please make it clear that what's the top-level errors look like. Will we keep the `InitError`, `ServiceError` and `StorageError`?

File: rfcs/47-better-error-handing.md [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+	}
+	```
+2. For the wrapped error:
+   - If it is an unexpected error, it MUST be `fmt.Errorf("%v", err)`. SDK errors SHOULD not be wrapped.
+   - If it is an expected error, it MUST be created by either
+     - `fmt.Errorf("%w: %v", SomeError, err)` where
+       - `SomeError` is defined as a sentinel error `var SomeError = errors.New("what happened")`
+       - `err` is the original SDK error
+     - `NewSomeError(contextA, contextB)` where 
+       - `SomeError` is returned and it is defined as:
```
Comment (Xuanwo) [inline rfcs/47-better-error-handing.md:119]: Can you add a real example here?

File: rfcs/47-better-error-handing.md [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+		ContextA string
+		ContextB structB
+		...
+	}
+	```
+2. For the wrapped error:
+   - If it is an unexpected error, it MUST be `fmt.Errorf("%v", err)`. SDK errors SHOULD not be wrapped.
+   - If it is an expected error, it MUST be created by either
+     - `fmt.Errorf("%w: %v", SomeError, err)` where
+       - `SomeError` is defined as a sentinel error `var SomeError = errors.New("what happened")`
```
Comment (Xuanwo) [inline rfcs/47-better-error-handing.md:116]: Please give a definition of `sentinel error`.

File: rfcs/47-better-error-handing.md [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+       - `SomeError` CAN implement `Unwrap`, returning the category (or the label) of the error, which is a sentinel error
+
+![](https://user-images.githubusercontent.com/37948597/116872935-6dd1fb00-ac49-11eb-8f25-8983d564ba66.png)
+
+In short, compared with current practice, there are three changes:
+1. Ad-hoc string errors SHOULD not be used.
+2. SDK errors SHOULD not be wrapped.
+3. Wrapped error `struct`s won't wrap another layer of error `struct`s, and its wrapped `Err` is hidden.
+
+Besides, we should list all available errors in the docs.
```
Comment (Xuanwo) [inline rfcs/47-better-error-handing.md:137]: The errors list should be introduced in this RFC or another one?

**Comment by xxchan** [comment]
I found I missed a (maybe important) fact that there's a classification of whether the error is caused inside `go-storage`, and interestingly the classification matches the classification based on error type definitions 🤔 
image

Let me think again.

**Comment by xxchan** [comment]
## A summary of the classification of errors by the cause of the error
- Errors which are only meaningful inside `go-storage`:
  - `go-storage` lacks capability: `ErrCapabilityInsufficient`
  - The user uses `go-storage` wrongly `ErrRestrictionDissatisfied`
- Errors which may appear when you use REST APIs with arbitrary clients:
  - Failed to send an HTTP request, e.g., invalid method, invalid URL. (We don't need to consider this?)
  - Failed to receive an HTTP response, e.g., socket timeouts and TCP disconnects.
  - Received an invalid HTTP response, e.g., missing Location header. (We don't need to consider this?)
  - Received an HTTP response with a non-200 status code, which may contain API error codes in the response body.

The so-called SDK errors, defined in vendors' docs, are just the last type. 

To be more clear, vendors' define the error responses of REST APIs, including the HTTP status code in the header and the error code in the body. Actually, it is the error code that matters. 

Then the SDKs may use this structure (status code + error code) directly or may provide some error definitions in their own way (which is what we have to figure out!).

**Comment by Xuanwo** [comment]
> Then the SDKs may use this structure (status code + error code) directly or may provide some error definitions in their own way (which is what we have to figure out!).

go-storage is not designed for HTTP only service, please also take local fs and services like hdfs / ceph (in which we use rpc or cgo to call service) into consideration.

**Comment by xxchan** [comment]
I think I should ignore the service implementations and focus on `go-storage`'s abstraction first:

We have `Object` and `Storager`. `Storager` is like a client, communicating with the storage system (HTTP server/RPC server/file system/...) to access `Object`.

Cases are:
1. `Storager` failed to send a request to the storage system
    1. incapability
    2. wrong usage (either `go-storage` or the specific service)
2. `Storager` failed to receive a response from the storage system
3. `Storager` received an invalid response from the storage system
4. `Storager` received an error response from the storage system

BTW I think this classification won't affect this RFC... 😕

**Comment by Xuanwo** [comment]
Maybe we can classify by the root cause of error?

- User Side: incapability, wrong usage, pass an append object into multipart operations (like  `failed to send a request to the storage system` )
- Network Side: Network related error (like `failed to receive a response from the storage system`)
- Service Side: not exist, permission denied, already exists... (like `received an error response from the storage system`)

As we can't handle the error caused by `invalid response from the storage system`, maybe we can return them in `fmt.Errorf("%v", err)`?

**Comment by xxchan** [comment]
So it will look like this
image

**Comment by Xuanwo** [comment]
LGTM

**Comment by xxchan** [comment]
The alternatives are not discussed:
1. Whether `InitError`, `ServiceError` & `StorageError` are actually needed?
2. Whether to warp none or all original errors?

**Comment by Xuanwo** [comment]
> Whether InitError, ServiceError & StorageError are actually needed?

I have no idea for that, maybe provide an example to show the differences?

> Whether to warp none or all original errors?

Avoid abstraction leak is a good catch, I think we should not wrap error from SDK. Maybe we can add a `UnknownError` for those error that go-storage can't handle?

**Comment by xxchan** [comment]
And which style is preferred?
## original
```go
// NewPairUnsupportedError will create a new PairUnsupportedError.
func NewPairUnsupportedError(pair types.Pair) *PairUnsupportedError {
	return &PairUnsupportedError{
		Err:  ErrCapabilityInsufficient,
		Pair: pair,
	}
}

// PairUnsupportedError means this operation has unsupported pair.
type PairUnsupportedError struct {
	Err error

	Pair types.Pair
}

func (e *PairUnsupportedError) Error() string {
	return fmt.Sprintf("pair unsupported, %s: %s", e.Pair, e.Err.Error())
}

// Unwrap implements xerrors.Wrapper
func (e *PairUnsupportedError) Unwrap() error {
	return e.Err
}
```


## hide `Err`
```go

// NewPairUnsupportedError will create a new PairUnsupportedError.
func NewPairUnsupportedError(pair types.Pair) *PairUnsupportedError {
	return &PairUnsupportedError{
		err:  ErrCapabilityInsufficient,
		Pair: pair,
	}
}

// PairUnsupportedError means this operation has unsupported pair.
type PairUnsupportedError struct {
	err error

	Pair types.Pair
}

func (e *PairUnsupportedError) Error() string {
	return fmt.Sprintf("pair unsupported, %s: %s", e.Pair, e.err.Error())
}

// Unwrap implements xerrors.Wrapper
func (e *PairUnsupportedError) Unwrap() error {
	return e.err
}
```

## remove `Err`
```go

// NewPairUnsupportedError will create a new PairUnsupportedError.
func NewPairUnsupportedError(pair types.Pair) *PairUnsupportedError {
	return &PairUnsupportedError{
		Pair: pair,
	}
}

// PairUnsupportedError means this operation has unsupported pair.
type PairUnsupportedError struct {
	Pair types.Pair
}

func (e *PairUnsupportedError) Error() string {
	return fmt.Sprintf("pair unsupported, %s: %s", e.Pair, ErrCapabilityInsufficient.Error())
}

// Unwrap implements xerrors.Wrapper
func (e *PairUnsupportedError) Unwrap() error {
	return ErrCapabilityInsufficient
}
```

**Comment by xxchan** [comment]
> > Whether InitError, ServiceError & StorageError are actually needed?
> 
> I have no idea for that, maybe provide an example to show the differences?

Oh, I just found they contain different context information, so it may be better to keep them

**Comment by Xuanwo** [comment]
I prefer the `remove err` style.

And can you show me an example that if the user calls `Stat(abc)` and got an ObjectNotExist error.

- what error struct he will get?
- How does he check the error?
- How does he got the related context details?

**Comment by xxchan** [comment]
```go
err = &services.StorageError{
	Op:       "stat",
	Err:      services.ErrObjectNotExist,
	Storager: s,
	Path:     "abc",
}
```
he can check `err.Is(err, services.ErrObjectNotExist)` 

`ErrObjectNotExist` has no related context. If he wants to get `Path`, he must 
```go
var sErr *services.StorageError
if errors.As(err, &sErr) {
	sErr.Path
}
```
🤔

**Comment by xxchan** [comment]
Emm, `ErrObjectNotExist` and `PairRequiredError` belong to the same level, but one is a value and the other is a type (and can be used in `errors.As`) 🤔

File: rfcs/47-better-error-handing.md (hunk 5 above)
Comment (xxchan) [reply rfcs/47-better-error-handing.md:177]: In the website doc page?

Comment (Xuanwo) [reply rfcs/47-better-error-handing.md:177]: In this RFC maybe?

Like global pairs, all errors in go-storage should be defined via the RFC procedure.

**Comment by xxchan** [comment]
demo:
https://github.com/aos-dev/go-storage/pull/555
https://github.com/aos-dev/go-service-s3/pull/78

File: rfcs/47-better-error-handing.md [hunk 6]
```diff
@@ -0,0 +1,251 @@
+---
+author: xxchan <xxchan22f@gmail.com>
+status: draft
+updated_at: 2020-05-03
+---
+
+# AOS-47: Better Error Handling
```
Comment (Xuanwo) [inline rfcs/47-better-error-handing.md:7]: We need a better proposal name here, a title that describes what has been changed in this proposal.

File: rfcs/47-better-error-handing.md [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+    - sentinel errors wrapped by `fmt.Errorf` with `%w`: 
+    	- defined in either [go-storage/services/error.go] or `go-service-*/error.go`
+    	- defined as `var ErrSomethingWrong = errors.New("what happened")`
+    	- returned in `fmt.Errorf("%w: %v", services.ErrObjectNotExist, err)`, carrying error messages of original SDK errors
+    - another layer of error `struct`s carrying contextual information: `MetadataUnrecognizedError`, `PairUnsupportedError`, `PairRequiredError`
+    	- defined in [go-storage/services/error.go]
+    	- have `Err` and context fields, don't have `Op`
+    	- returned by constructor methods, where `Err` is set to a sentinel error, which can be viewed as the error `struct`s' classification
+
+![](./47/old.png)
```
Comment (Xuanwo) [inline rfcs/47-better-error-handing.md:55]: Let's add an empty line under images.

File: rfcs/47-better-error-handing.md [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+
+	return err
+}
+```
+
+Although we declared in `11-error-handling` that unexpected errors could be changed or disappeared while dependence upgraded and no changelog for them, we'd better simply eliminate the possibility of using them.
+
+#### Expected Errors
+
+`11-error-handling` is not detailed enough. It does not specify how to define and use expected errors. Current practice has too many cases, and has some problems. Both users and implementors may be confused which kind of error should be returned at which place.
```
Comment (Xuanwo) [inline rfcs/47-better-error-handing.md:86]: Use `AOS-11` to refer `11-error-handling` instead.

File: rfcs/47-better-error-handing.md [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+		ContextA string
+		ContextB structB
+		...
+	}
+	```
+   - SHOULD implement `Error` and `Unwrap`
+   - Only include `InitError`, `ServiceError` and `StorageError`
+2. For the wrapped error:
+   - If it is an unexpected error, it MUST be `fmt.Errorf("%w: %v", ErrUnexpected, err)`
+     - `ErrUnexpected` is defined as: `var ErrUnexpected = errors.New("go-storage can't handle this error")`.
```
Comment (Xuanwo) [inline rfcs/47-better-error-handing.md:116]: `go-storage can't handle this error` is too verbose for this error, how about `unexpected`?

File: rfcs/47-better-error-handing.md [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+    - Top-level errors decide which component the errors belong to.
+    - Wrapped errors are specific. There isn't a more specific error type.
+    - If some wrapped errors can be grouped together, their `Unwrap` will return a sentinel error representing the category.
+- Provide a unified user experience: define an abstract layer of errors for the users, free them of the tedium of handling similar errors from multiple SDKs.
+
+## Compatibility
+
+The following changes will be made:
+- Extract some ad-hoc string errors into defined sentinel errors. (forward compatible)
+- Turn SDK errors into `fmt.Errorf("%v", err)`. (not forward compatible, but doesn't violate our promise)
```
Comment (Xuanwo) [inline rfcs/47-better-error-handing.md:243]: Use `ErrUnexpected` instead?

File: rfcs/47-better-error-handing.md [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+  - If we have too many levels, the users may use too general or too specific errors. We should provide proper granularity. The proposal's granularity can be described as:
+    - Top-level errors decide which component the errors belong to.
+    - Wrapped errors are specific. There isn't a more specific error type.
+    - If some wrapped errors can be grouped together, their `Unwrap` will return a sentinel error representing the category.
+- Provide a unified user experience: define an abstract layer of errors for the users, free them of the tedium of handling similar errors from multiple SDKs.
+
+## Compatibility
+
+The following changes will be made:
+- Extract some ad-hoc string errors into defined sentinel errors. (forward compatible)
```
Comment (Xuanwo) [inline rfcs/47-better-error-handing.md:242]: Please list in `Implementation` sections.

File: rfcs/47-better-error-handing.md [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+- Turn SDK errors into `fmt.Errorf("%v", err)`. (not forward compatible, but doesn't violate our promise)
+- Some error `struct`s' `Err` field will be removed. (not forward compatible)
+
+## Implementation
+
+Most of the work would be done by the author of this proposal.
```
Comment (Xuanwo) [inline rfcs/47-better-error-handing.md:248]: Please detail the works here and convert them into an issue after this proposal has been approved.

File: rfcs/47-better-error-handing.md [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+
+### All currently available global expected errors 
+
+- ErrCapabilityInsufficient
+  - MetadataUnrecognizedError
+  - PairUnsupportedError
+- ErrRestrictionDissatisfied
+  - PairRequiredError
+- ErrObjectNotExist
+- ErrObjectModeInvalid
```
Comment (Xuanwo) [inline rfcs/47-better-error-handing.md:210]: Do we need to introduce a new error struct that specifies `expected mode` and `actual mode`?

File: rfcs/47-better-error-handing.md [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+- ErrPermissionDenied
+- ErrListModeInvalid
```
Comment (Xuanwo) [inline rfcs/47-better-error-handing.md:212]: same question.

File: rfcs/47-better-error-handing.md [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+- `ObjetNotExist`
+- `ObjectModeInvalid`
+
+instead of
+- `NotExistObject`
+- `InvalidObjectMode`
+
+### All currently available global expected errors 
+
+- ErrCapabilityInsufficient
```
Comment (Xuanwo) [inline rfcs/47-better-error-handing.md:204]: Please design an error code for https://github.com/aos-dev/go-storage/blob/master/services/new.go#L61

File: rfcs/47-better-error-handing.md [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+  - MetadataUnrecognizedError
+  - PairUnsupportedError
+- ErrRestrictionDissatisfied
+  - PairRequiredError
+- ErrObjectNotExist
+- ErrObjectModeInvalid
+- ErrPermissionDenied
+- ErrListModeInvalid
+
```
Comment (Xuanwo) [inline rfcs/47-better-error-handing.md:213]: The network-related error will be introduced in another RFC? Like `io.EOF` and `context.Canceled`?

File: rfcs/47-better-error-handing.md (hunk 6 above)
Comment (xxchan) [reply rfcs/47-better-error-handing.md:7]: What about: Additional error specification

Comment (Xuanwo) [reply rfcs/47-better-error-handing.md:7]: LGTM

File: rfcs/47-better-error-handing.md (hunk 13 above)
Comment (xxchan) [reply rfcs/47-better-error-handing.md:210]: If an operation needs a specific mode, then `expected mode` is fine.
If an operation allows a set of different modes, how to handle this case?

File: rfcs/47-better-error-handing.md (hunk 16 above)
Comment (xxchan) [reply rfcs/47-better-error-handing.md:213]: I think it requires discussion and is worth a separate RFC

Comment (Xuanwo) [reply rfcs/47-better-error-handing.md:213]: OK

File: rfcs/47-better-error-handing.md (hunk 13 above)
Comment (Xuanwo) [reply rfcs/47-better-error-handing.md]: Object mode's design makes sure that object mode is orthogonal, so an operation that requires object mode check only allows a single mode here.

- WriteAppend requires `IsAppend`
- ListPart requires `IsPart`
- Read requires `IsRead` (not implemented, added in https://github.com/aos-dev/go-storage/issues/557)
- Write doesn't need object mode check

Comment (xxchan) [reply rfcs/47-better-error-handing.md]: What about this:
https://github.com/aos-dev/go-service-s3/blob/84e8d53888817a5508aaf54f260a8e9d1a7fb37b/storage.go#L173

Comment (Xuanwo) [reply rfcs/47-better-error-handing.md]: `ListMode` is also orthogonal, but I think return `actual mode` is enough for `ListModeInvalid` error.

**Comment by xxchan** [comment]
I think the `New` function here can be removed since we remove the `Err` field (but this is break change, or we can also keep existing ones but do not create ones for new error `struct`s)
```go
// NewPairUnsupportedError will create a new PairUnsupportedError.
func NewPairUnsupportedError(pair types.Pair) *PairUnsupportedError {
	return &PairUnsupportedError{
		Pair: pair,
	}
}

// PairUnsupportedError means this operation has unsupported pair.
type PairUnsupportedError struct {
	Pair types.Pair
}

func (e *PairUnsupportedError) Error() string {
	return fmt.Sprintf("pair unsupported, %s: %s", e.Pair, ErrCapabilityInsufficient.Error())
}

// Unwrap implements xerrors.Wrapper
func (e *PairUnsupportedError) Unwrap() error {
	return ErrCapabilityInsufficient
}
```

**Comment by Xuanwo** [comment]
PairUnsupportedError only used in generated code, I think it's OK to remove the `New` function.

BTW, let's use struct here to avoid extra heap alloc:

```go
func (e PairUnsupportedError) Error() string {
	return fmt.Sprintf("pair unsupported, %s: %s", e.Pair, ErrCapabilityInsufficient.Error())
}

// Unwrap implements xerrors.Wrapper
func (e PairUnsupportedError) Unwrap() error {
	return ErrCapabilityInsufficient
}
```

**Comment by xxchan** [comment]
I begin to wonder why `Error` is usually implemented with pointer receiver 🤔

File: rfcs/47-additional-error-specification.md [hunk 17]
```diff
@@ -206,10 +206,11 @@ instead of
   - PairUnsupportedError
 - ErrRestrictionDissatisfied
   - PairRequiredError
+- ObjectModeInvalidError
+- ListModeInvalidError
 - ErrObjectNotExist
-- ErrObjectModeInvalid
 - ErrPermissionDenied
-- ErrListModeInvalid
+- ErrNewFuncNotRegistered
```
Comment (Xuanwo) [inline rfcs/47-additional-error-specification.md:213]: I prefer `ErrServiceNotRegistered`

**Comment by xxchan** [comment]
Will it be a problem if pointer and value receiver are mixed? e.g., user will need to:
```go
var errA *ErrorA
errors.As(err, &errA)

var errB ErrorB
errors.As(err. &errB)
```

**Comment by Xuanwo** [comment]
> ```go
> var errA *ErrorA
> errors.As(err, &errA)
> 
> var errB ErrorB
> errors.As(err. &errB)
> ```

Yes, we should stick to the same style here.

**Comment by Xuanwo** [comment]
This spec is OK to merge!

File: rfcs/47-additional-error-specification.md [hunk 18]
```diff
@@ -0,0 +1,291 @@
+---
+author: xxchan <xxchan22f@gmail.com>
+status: draft
+updated_at: 2020-05-03
```
Comment (Xuanwo) [inline rfcs/47-additional-error-specification.md:4]: Update `updated_at`

**Comment by xxchan** [comment]
> This spec is OK to merge!

Wait for demo updated, I think

**Comment by xxchan** [comment]
I think I found a reason why pointer receiver can be better...
If `Error` is implemented for `ErrA`, then both `ErrA` and `*ErrA` can be used as `error`. And they are two different error types. So when using `As`, confusion may appear... Think of `ErrA` should be returned, but somewhere `*ErrA` is returned by mistake.
https://play.golang.org/p/hBltJt8ulkV

**Comment by Xuanwo** [comment]
> I think I found a reason why pointer receiver can be better...
> If `Error` is implemented for `ErrA`, then both `ErrA` and `*ErrA` can be used as `error`. And they are two different error types. So when using `As`, confusion may appear... Think of `ErrA` should be returned, but somewhere `*ErrA` is returned by mistake.
> https://play.golang.org/p/hBltJt8ulkV

I can't open the <https://play.golang.org/p/hBltJt8ulkV> for `Unavailable For Legal Reasons` but I can guess why and I think it's correct.

So we need to make sure that all errors returned by struct value instead of pointer.

If needed, we can implement this check in go-storage code generator.

**Comment by xxchan** [comment]
Oh, check in codegen might be a good idea. 

But...If using pointer receiver is common practice, and using value receiver requires an additional check, why do you prefer value receiver then? 
And will the added copy & check overheads outweigh reduced allocation overheads (I have no idea about this performance issue)

**Comment by Xuanwo** [comment]
> Oh, check in codegen might be a good idea.
> 
> But...If using pointer receiver is common practice, and using value receiver requires an additional check, why do you prefer value receiver then?
> And will the added copy & check overheads outweigh reduced allocation overheads (I have no idea about this performance issue)

For now, there is no plan to implement the check in codegen, we only make it as a promise: all errors returned by go-storage and go-services are struct value instead of pointers. And the check will be finished at build time instead of runtime (If we found service provider return error via pointer, we will print error during `make build`).

So why I prefer to return struct value instead of pointer value? Let's do some benchmarks.

```go
import "testing"

const structSize = 100

type Struct struct {
	v [structSize]int
}

//go:noinline
func returnValue() Struct {
	var s Struct
	for i := 0; i < structSize; i++ {
		s.v[i] = i
	}
	return s
}

//go:noinline
func returnPointer() *Struct {
	var s Struct
	for i := 0; i < structSize; i++ {
		s.v[i] = i
	}
	return &s
}

func BenchmarkStructValue(b *testing.B) {
	b.ReportAllocs()

	t := 0
	for i := 0; i < b.N; i++ {
		v := returnValue()
		t += v.v[0]
	}
}

func BenchmarkPointerValue(b *testing.B) {
	b.ReportAllocs()

	t := 0
	for i := 0; i < b.N; i++ {
		v := returnPointer()
		t += v.v[0]
	}
}
```

Here:

- Added `//go:noinline` to make the function is not inlined by golang (like our user logic, they are often too complicated to be inlined)
- Added `b.ReportAllocs()` to let golang report allocs
- Create a struct will a specific size (golang's struct memory is a bit like C, so we can use `[10]int` instead of ten int field)

So while the struct size like the following, we can get results like:

1024 (already a big struct size)

```
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkStructValue
BenchmarkStructValue-8    	 1640901	       686.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkPointerValue
BenchmarkPointerValue-8   	  548283	      1937 ns/op	    8192 B/op	       1 allocs/op
PASS
```

1024 * 64 (a struct with 64K int, very huge struct)

```
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkStructValue
BenchmarkStructValue-8    	   14569	     78709 ns/op	       0 B/op	       0 allocs/op
BenchmarkPointerValue
BenchmarkPointerValue-8   	    9802	    111656 ns/op	  524289 B/op	       1 allocs/op
PASS
```

1024 * 1024(a struct with 1M int, embed a while file in it)

```
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkStructValue
BenchmarkStructValue-8    	     270	   3733077 ns/op	       0 B/op	       0 allocs/op
BenchmarkPointerValue
BenchmarkPointerValue-8   	     464	   2321941 ns/op	 8388617 B/op	       1 allocs/op
PASS
```

This is not a very strict performance on struct value and pointer value, but we can still outline some points:

- returning struct value usually faster than pointer value and don't need extra allocs, save memory space and reduce the pressure on Golang GC
- values that live shortly is better to use struct value
- values that could be used often is better to use the pointer value

So, I prefer to return a struct value as error instead.

**Comment by xxchan** [comment]
A good illustration of allocating is expensive! 
I agree it is better to use struct value for values that live shortly, and errors are short-lived. (But errors are seldom returned, so the performance issue is not very important?) 
Anyway, let's use value receiver.

**Comment by xxchan** [comment]
ready now  

rendered ver: https://github.com/xxchan/specs/blob/error/rfcs/47-additional-error-specification.md

demo
https://github.com/aos-dev/go-storage/pull/555
https://github.com/aos-dev/go-service-s3/pull/78

File: rfcs/47-additional-error-specification.md [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+
+// Unwrap implements xerrors.Wrapper
+func (e PairUnsupportedError) Unwrap() error {
+	return ErrCapabilityInsufficient
+}
+```
+
+![](./47/new.png)
+
+In short, compared with current practice, there are three changes:
```
Comment (Xuanwo) [inline rfcs/47-additional-error-specification.md:171]: Add an empty line under this line.

File: rfcs/47-additional-error-specification.md [hunk 20]
```diff
@@ -0,0 +1,297 @@
+---
+author: xxchan <xxchan22f@gmail.com>
+status: draft
+updated_at: 2020-05-07
```
Comment (Prnyself) [inline rfcs/47-additional-error-specification.md:4]: 2021-05-07

Comment (xxchan) [reply rfcs/47-additional-error-specification.md:4]: 😅 😭

File: rfcs/47-additional-error-specification.md [hunk 21]
```diff
@@ -0,0 +1,297 @@
...
+    - Will they wrap yet another layer of error `struct`s? 
+    - If it will be the same for all instances, not intended for the caller to fill in, and only accessed via `Unwrap`, we can actually hide the field.
+    - The wrapped errors seem like the super class. It may be more reasonable to wrap other errors instead of being wrapped. But the order does not matter in the error chain.
+
+## Proposal
+
+So I propose the following error handling specification as a supplement of [AOS-11]:
+
+An error code is an exported public variable containing an `error` value created by `errors.New()` (a.k.a. a sentinel error)
+- `var SomeError = errors. New("what happened")`
```
Comment (Prnyself) [inline rfcs/47-additional-error-specification.md:104]: Remove extra space in `errors.New`

File: rfcs/47-additional-error-specification.md [hunk 22]
```diff
@@ -0,0 +1,297 @@
...
+			type SomeError struct {
+				ContextA string
+				ContextB structB
+				...
+			}
+			```
+       - it SHOULD implement `Unwrap`, returning its error code
+       - it SHOULD implement `Error`
+         - the format should be `{Description}, {ContextA}, {ContextB}: {Err}` , where `Err` is its error code
+         - or `{Description}, {ContextA}, {ContextB}` if it is the only struct with its error code
```
Comment (Prnyself) [inline rfcs/47-additional-error-specification.md:137]: `with` or `without`?

Comment (xxchan) [reply rfcs/47-additional-error-specification.md:137]: `with`, I mean the error code and the struct is one-one relation, e.g. `ErrObjectModeInvalid` -  `ObjectModeInvalidError`

Now it's `object mode invalid, expected %v, actual %v` 
Otherwise it will be `object mode invalid, expected %v, actual %v: %s`
i.e.,  `object mode invalid, expected %v, actual %v: object mode invalid`

I think it's too wordy?

Comment (xxchan) [reply rfcs/47-additional-error-specification.md:137]: And there's no struct without an error code now

**Review by Xuanwo — APPROVED** [review]
